### PR TITLE
docs(boot): Docs for @loopback/boot

### PR DESCRIPTION
**Pull request by virkt25**
Documentation explaining the `@loopback/boot` extension and the various aspects that make it up.

Need a review and suggestions on the following:
-- Should individual Booters be documented here? (Or link to docs in README for the package?)
-- Layout?
-- Grammar?

### CODE REVIEW

File: pages/en/lb4/Booting-an-Application.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+A typical LoopBack application is made up of many artifacts in different files,
+organized in different folders. **Booting an Application** means:
+
+* Discover artifacts automatically based on a convention (a specific folder
+  containing files with a given suffix)
+* Process those artifacts automatically (this usually means binding them to the Application)
+
+`@loopback/boot` provides a Bootstrapper that uses Booters to automatically
+discover and bind artifacts, all packaged in an easy to use Mixin.
```
Comment (dhmlau) [inline pages/en/lb4/Booting-an-Application.md]: "easy to use" => "easy-to-use"?

File: pages/en/lb4/Booting-an-Application.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+### What is an artifact?
+
+An artifact is any LoopBack construct usually defined in code as a Class. The word
+artifact describes Controllers, Repositories, Models, etc.
```
Comment (dhmlau) [inline pages/en/lb4/Booting-an-Application.md]: Suggestion to reword the 2nd sentence:
LoopBack constructs include Controllers, Repositories.....

File: pages/en/lb4/Booting-an-Application.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  app.projectRoot = __dirname;
+  await app.boot();
+  await app.start();
+}
+```
+
+### app.booters()
+
+A convenience method manually bind `Booters`. You can pass any number of `Booter`
+classes to this method and they will all get bound to the Application using the
```
Comment (dhmlau) [inline pages/en/lb4/Booting-an-Application.md]: get bound => get bounded

File: pages/en/lb4/Booting-an-Application.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+```ts
+// Binds MyCustomBooter to `booters.MyCustomBooter`
+// Binds AnotherCustomBooter to `booters.AnotherCustomBooter`
+// Both will have the `booter` tag set.
+app.booters(MyCustomBooter, AnotherCustomBooter);
+```
+
+## BootComponent
+
+This component is added to a Application by `BootMixin` if used. The Component:
```
Comment (dhmlau) [inline pages/en/lb4/Booting-an-Application.md]: "a Application" => "an Application"? 
"The Component" => "This Component"

File: pages/en/lb4/Booting-an-Application.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+bind the `booters` of a component manually._
+
+```ts
+app.component(BootComponent);
+```
+
+## Bootstrapper
+
+A Class that acts as the "manager" for Booters. The Boostrapper is designed to be
+bound to an Application as a `SINGLETON`. The Bootstrapper class provides a `boot()`
```
Comment (dhmlau) [inline pages/en/lb4/Booting-an-Application.md]: bound -> bounded

File: pages/en/lb4/Booting-an-Application.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+method. This method is responsible for getting all bound `Booters` and running
+their `phases`. A `phase` is a method on a `Booter` class.
+
+Each call of the `boot()` method creates a new `Context` that sets the `app` context
```
Comment (dhmlau) [inline pages/en/lb4/Booting-an-Application.md]: suggestion:
"Each call of the `boot()` method" => "Each `boot()` method call"

File: pages/en/lb4/Booting-an-Application.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+as it's parent. This is done so each `Context` for `boot` gets a new instance of
```
Comment (dhmlau) [inline pages/en/lb4/Booting-an-Application.md]: it's parent => its parent

**Review by dhmlau — COMMENTED** [review]
While this doc might have good information on what boot is and how it works, I might miss the points (as a user):
- why do i need to care about how boot if artifacts are automatically loaded for me when using CLI
- in what case i need to know more about booters

File: pages/en/lb4/Booting-an-Application.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+discover and bind artifacts, all packaged in an easy to use Mixin.
+
+### What is an artifact?
+
+An artifact is any LoopBack construct usually defined in code as a Class. The word
+artifact describes Controllers, Repositories, Models, etc.
+
+## Usage: @loopback/cli
+
+New projects generated using `@loopback/cli` or `lb4` are automatically enabled
```
Comment (dhmlau) [inline pages/en/lb4/Booting-an-Application.md]: I wonder if it's a good idea to let the readers know up front that for general use, artifacts are automatically loaded for you and you don't need to know too much details.

File: pages/en/lb4/Booting-an-Application.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+Since this is a convention based Bootstrapper, it is important to set a `projectRoot`,
+as all other artifact paths will be resolved relative to this path.
+
+_Tip_: `application.ts` will likely be at the root of your project, so it's path can be
+used to set the `projectRoot` by using the `__dirname` variable. _(See example below)_
+
+#### Using the BootMixin
+
+`Booter` and `Binding` types must be imported alongside `BootMixin` to allow TypeScript
+to infer types and avoid errors. \_If using `tslint` with the `no-unused-variable` rule,
```
Comment (shimks) [inline pages/en/lb4/Booting-an-Application.md]: `\_`

File: pages/en/lb4/Concepts.md [hunk 10]
```diff
@@ -10,22 +10,24 @@ summary:
...
-- [**Repository**](Repositories.html): Type of Service that represents a collection of data within a DataSource.
-- [**Decorator**](Decorators.html): Enables you to annotate or modify your class declarations and members with metadata.
-- **Component**: A package that bundles one or more Loopback extensions.
-  - See [Using components](Using-components.html) and [Creating components](Creating-components.html) for more information.
+* [**Application**](Application.html): In LoopBack 4, the Application class is
+  the central class for setting up all of your module’s components, controllers,
+  servers and bindings. The Application class extends [Context](Context.html), and
+  provides the controls for starting and stopping itself and its associated
+  servers.
+* [**Booting an Application**](Booting-an-Application.html): A convention based
```
Comment (shimks) [inline pages/en/lb4/Concepts.md]: I think the section on `boot` should be somewhere at the bottom if the order matters. Most users of LoopBack won't really have to know about the fine details of booters other than that they do context bindings for you. I think it'd be more appropriate below `Context` here.

File: pages/en/lb4/Booting-an-Application.md [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+---
+
+## What does it mean to boot an Application?
+
+A typical LoopBack application is made up of many artifacts in different files,
+organized in different folders. **Booting an Application** means:
+
+* Discover artifacts automatically based on a convention (a specific folder
+  containing files with a given suffix)
+* Process those artifacts automatically (this usually means binding them to the Application)
```
Comment (shimks) [inline pages/en/lb4/Booting-an-Application.md]: I think we should use fit in the concept of `context` in here somewhere, just to make sure that users understand that `boot` takes care of automatic **`context`** bindings.

**Review by shimks — COMMENTED** [review]
Nice docs. It's missing a sidebar entry and there are some minor nitpicks, but once that's been taken care of, LGTM.

File: pages/en/lb4/Booting-an-Application.md (hunk 3 above)
Comment (virkt25) [reply pages/en/lb4/Booting-an-Application.md]: Is `be bound` better?

**Comment by virkt25** [comment]
@dhmlau Good points! Thoughts on splitting this doc into 2 different pages? 
- One for "users" of the Application (all they need to know is CLI projects are enables, how to add to manual application and the config options avaiable to them) 
- Another for Extension Developers who would have more of an interest in understanding how `@loopback/boot` works and how they can leverage this knowledge to write their own `Booter` -- realistically I think an extension developer just needs to know how to write a Booter which doesn't require details about the inner workings

Does anyone need to know how Boot works under the hood?

**Comment by dhmlau** [comment]
+1 on splitting into 2 pages or at least 2 sections. 
I wonder if the split is MVP.  But the other argument is we might as well get it all fixed at one time, rather than going back again when we have time.

File: pages/en/lb4/Booting-an-Application.md [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+Boot functionality can be added to a LoopBack 4 Application by mixing it with the
+`BootMixin`. The Mixin adds the `BootComponent` to your Application as well as
+convenience methods such as `app.boot()` and `app.booters()`. The Mixin also allows
+Components to set the property `booters` as an Array of `Booters`. They will be bound
+to the Application and called by the `Bootstrapper`.
+
+Since this is a convention based Bootstrapper, it is important to set a `projectRoot`,
+as all other artifact paths will be resolved relative to this path.
+
+_Tip_: `application.ts` will likely be at the root of your project, so it's path can be
```
Comment (kjdelisle) [inline pages/en/lb4/Booting-an-Application.md]: `it's` -> `its`

File: pages/en/lb4/Booting-an-Application.md [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+
+## BootMixin
+
+Boot functionality can be added to a LoopBack 4 Application by mixing it with the
+`BootMixin`. The Mixin adds the `BootComponent` to your Application as well as
+convenience methods such as `app.boot()` and `app.booters()`. The Mixin also allows
+Components to set the property `booters` as an Array of `Booters`. They will be bound
+to the Application and called by the `Bootstrapper`.
+
+Since this is a convention based Bootstrapper, it is important to set a `projectRoot`,
```
Comment (kjdelisle) [inline pages/en/lb4/Booting-an-Application.md]: `convention-based`

File: pages/en/lb4/Booting-an-Application.md [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+    // Setting the projectRoot
+    this.projectRoot = __dirname;
+  }
+}
+```
+
+### app.boot()
+
+A convenience method to retrieve the `Bootstrapper` instance bound to the
+Application and calls it's `boot` function. This should be called before an
```
Comment (kjdelisle) [inline pages/en/lb4/Booting-an-Application.md]: `it's` -> `its`

File: pages/en/lb4/Booting-an-Application.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+  const app = new MyApp();
+  app.projectRoot = __dirname;
+  await app.boot();
+  await app.start();
+}
+```
+
+### app.booters()
+
+A convenience method manually bind `Booters`. You can pass any number of `Booter`
```
Comment (kjdelisle) [inline pages/en/lb4/Booting-an-Application.md]: `A convenience method to manually bind...`

File: pages/en/lb4/Booting-an-Application.md [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+classes to this method and they will all get bounded to the Application using the
```
Comment (kjdelisle) [inline pages/en/lb4/Booting-an-Application.md]: `..and they will all be bound...`

File: pages/en/lb4/Booting-an-Application.md [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+_If using this as a standalone component without the `BootMixin`, you will need to
+bind the `booters` of a component manually._
+
+```ts
+app.component(BootComponent);
+```
+
+## Bootstrapper
+
+A Class that acts as the "manager" for Booters. The Boostrapper is designed to be
```
Comment (kjdelisle) [inline pages/en/lb4/Booting-an-Application.md]: `Boostrapper` -> `Bootstrapper`

:P

File: pages/en/lb4/Booting-an-Application.md [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+bounded to an Application as a `SINGLETON`. The Bootstrapper class provides a `boot()`
```
Comment (kjdelisle) [inline pages/en/lb4/Booting-an-Application.md]: `bounded` -> `bound`

File: pages/en/lb4/Booting-an-Application.md [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+method. This method is responsible for getting all bound `Booters` and running
+their `phases`. A `phase` is a method on a `Booter` class.
+
+Each `boot()` method call creates a new `Context` that sets the `app` context
+as its parent. This is done so each `Context` for `boot` gets a new instance of
+`booters` but the same context can be passed into `boot` so selective `phases` can be
+run in different calls of `boot`.
+
+The boostrapper can be configured to run only certain booters or phases of booters
```
Comment (kjdelisle) [inline pages/en/lb4/Booting-an-Application.md]: `boostrapper` -> `bootstrapper`

Comment (kjdelisle) [reply pages/en/lb4/Booting-an-Application.md]: I'd also rephrase it a bit: `...can be configured to run specific booters, or boot phases...`

File: pages/en/lb4/Booting-an-Application.md [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+by passing in `BootExecOptions`. **This is experimental and subject to change. Hence,
+this functionality is not exposed when calling `boot()` via `BootMixin`**.
+
+To use `BootExecOptions` you must directly call `bootstrapper.boot()` instead of `app.boot()`.
```
Comment (kjdelisle) [inline pages/en/lb4/Booting-an-Application.md]: Need a comma after `BootExecOptions`.

"To use BootExecOptions, you must..."
(I can't use the backticks in the example, messes up the formatting.)

File: pages/en/lb4/Concepts.md [hunk 21]
```diff
@@ -10,22 +10,24 @@ summary:
...
+* [**Application**](Application.html): In LoopBack 4, the Application class is
+  the central class for setting up all of your module’s components, controllers,
+  servers and bindings. The Application class extends [Context](Context.html), and
+  provides the controls for starting and stopping itself and its associated
+  servers.
+* [**Server**](Server.html): Represents implementation for inbound transports and/or protocols such as REST over http, gRPC over http2, and graphQL over https. It typically listens for requests on a specific port, handle them, and return appropriate responses.
+* [**Context**](Context.html): An abstraction of all state and dependencies in your application, that LoopBack uses to “manage” everything. It's a global registry for everything in your app (configurations, state, dependencies, classes, and so on).
+* [**Dependency Injection**](Dependency-injection.html): Technique that separates the construction of dependencies of a class or function from its behavior, to keep the code loosely coupled.
+* [**Booting an Application**](Booting-an-Application.html): A convention based
```
Comment (kjdelisle) [inline pages/en/lb4/Concepts.md]: `convention-based`

**Review by kjdelisle — CHANGES_REQUESTED** [review]
Excellent overall, just has some minor spelling, grammar and phrasing problems to address and then it can land.

**Comment by dhmlau** [comment]
> +1 on splitting into 2 pages or at least 2 sections.
I wonder if the split is MVP. But the other argument is we might as well get it all fixed at one time, rather than going back again when we have time.

Please feel free to create separate issue to track that.  I don't think it should be a blocker to land this PR.

File: pages/en/lb4/Booting-an-Application.md [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+A Booter is a Class that implements the `Booter` interface. The Class
+must implement methods that corresponds to a `phase` name. The `phases` are called
+by the Bootstrapper in a pre-determined order (unless overridden by `BootExecOptions`).
+The next phase is only called once the previous phase has been completed for all Booters.
+
+### Phases
+
+#### configure
+
+Used to configure the `Booter` with it's default options.
```
Comment (b-admike) [inline pages/en/lb4/Booting-an-Application.md]: `it's` -> `its`

File: pages/en/lb4/Booting-an-Application.md [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+| Property         | Type                    | Description                                      |
+| ---------------- | ----------------------- | ------------------------------------------------ |
+| `booters`        | `Constructor<Booter>[]` | Array of Booters to bind before running `boot()` |
+| `filter.booters` | `string[]`              | Names of Booter classes that should be run       |
+| `filter.phases`  | `string[]`              | Names of Booter phases to run                    |
+
+### Example
+
+```ts
+class MyApp extends BootMixin(Application) {}
```
Comment (b-admike) [inline pages/en/lb4/Booting-an-Application.md]: can we show the imports here?

File: pages/en/lb4/Booting-an-Application.md [hunk 24]
```diff
@@ -0,0 +1,238 @@
...
+This Booter's purpose is to discover [Controller](http://loopback.io/doc/en/lb4/Controllers.html) type Artifacts and to bind
+them to the Application's Context.
+
+You can configure the conventions used in your
+project for a Controller by passing a `controllers` object on `BootOptions` property
+of your Application. The `controllers` object supports the following options:
+
+| Options      | Type                | Default              | Description                                                                                                   |
+| ------------ | ------------------- | -------------------- | ------------------------------------------------------------------------------------------------------------- |
+| `dirs`       | `string | string[]` | `['controllers']`    | Paths relative to projectRoot to look in for Controller artifacts                                             |
```
Comment (shimks) [inline pages/en/lb4/Booting-an-Application.md]: The table is broken here

File: pages/en/lb4/Booting-an-Application.md [hunk 25]
```diff
@@ -0,0 +1,238 @@
...
+them to the Application's Context. The use of this Booter requires `RepositoryMixin`
+from `@loopback/repository` to be mixed into your Application class.
+
+You can configure the conventions used in your
+project for a Repository by passing a `repositories` object on `BootOptions` property
+of your Application. The `repositories` object supports the following options:
+
+| Options      | Type                | Default              | Description                                                                                                   |
+| ------------ | ------------------- | -------------------- | ------------------------------------------------------------------------------------------------------------- |
+| `dirs`       | `string | string[]` | `['repositories']`   | Paths relative to projectRoot to look in for Repository artifacts                                             |
```
Comment (shimks) [inline pages/en/lb4/Booting-an-Application.md]: Broken here as well

**Review by shimks — APPROVED** [review]
Some tables are broken, but once that's been fixed, :shipit:

**Review by b-admike — APPROVED** [review]
+1 for Kyu's comments, LGTM otherwise

**Review by bschrammIBM — APPROVED** [review]

**Comment by virkt25** [comment]
Good catch @shimks. Used `|` in the table without escaping it. Weird that the backticks don't escape it automatically.

**Review by kjdelisle — APPROVED** [review]

**Pull request merged**